include/simics/dmllib.h

```diff
@@ -599,6 +599,422 @@ _identity_eq(const _identity_t a, const _identity_t b) {
+        return Sim_Set_Ok;
+    }
+}
+
+UNUSED static set_error_t
+_deserialize_simple_event_arguments(
+    size_t args_size,
+    set_error_t (*args_deserializer)(attr_value_t *in, void *out),
+    attr_value_t *arguments_attr, void **out_args) {
+
+    if (arguments_attr) {
+        if (!args_deserializer) {
+            if (!SIM_attr_is_list(*arguments_attr)
+                || SIM_attr_list_size(*arguments_attr) != 0) {
+                SIM_attribute_error(
+                    "Invalid serialized value of after event data: expected "
+                    "empty list as serialized value for arguments");
+                return Sim_Set_Illegal_Value;
+            }
+            *out_args = NULL;
+            return Sim_Set_Ok;
+        } else {
+            void *temp_out_args = MM_MALLOC(args_size, uint8);
+            set_error_t error = args_deserializer(arguments_attr,
+                                                  temp_out_args);
+            if (error != Sim_Set_Ok) {
+                MM_FREE(temp_out_args);
+                return error;
+            }
+            *out_args = temp_out_args;
+            return Sim_Set_Ok;
+        }
+    } else if (args_deserializer) {
+        SIM_attribute_error(
+            "Invalid serialized value of after event data: missing entry "
+            "for key 'arguments'");
+        return Sim_Set_Illegal_Value;
+    } else {
+        *out_args = NULL;
+        return Sim_Set_Ok;
+    }
+}
+
+UNUSED static set_error_t
+_deserialize_simple_event_domains(
+    ht_str_table_t *id_info_ht, attr_value_t *domains_attr,
+    _identity_t **out_domains, uint64 *out_no_domains) {
+
+    if (domains_attr) {
+        if (!SIM_attr_is_list(*domains_attr)) {
+            SIM_attribute_error(
+                "Invalid serialized representation of after event data: "
+                "specified domains is not a list.");
+            return Sim_Set_Illegal_Type;
+        }
+        uint32 temp_out_no_domains = SIM_attr_list_size(*domains_attr);
+        if (temp_out_no_domains == 0) {
+            *out_no_domains = 0;
+            *out_domains = NULL;
+            return Sim_Set_Ok;
+        }
+        _identity_t *temp_out_domains = MM_MALLOC(temp_out_no_domains,
+                                                  _identity_t);
+
+        attr_value_t *ids = SIM_attr_list(*domains_attr);
+        for (uint64 i = 0; i < temp_out_no_domains; ++i) {
+            set_error_t error = _deserialize_identity(id_info_ht, ids[i],
+                                                      &temp_out_domains[i]);
+            if (error != Sim_Set_Ok) {
+                MM_FREE(temp_out_domains);
+                return error;
+            }
+        }
+        *out_no_domains = temp_out_no_domains;
+        *out_domains = temp_out_domains;
+    } else {
+        *out_no_domains = 0;
+        *out_domains = NULL;
+    }
+    return Sim_Set_Ok;
+}
+
+UNUSED static set_error_t
+_deserialize_simple_event_data(
```

Comment:
It's kind of hilarious how about 50% of the entire PR is dedicated to the code necessary to implement this function.

Reply:
The functions all look nice and testable. It would be nice to have some unit testing; either as a pure C file in the project, or within a header/footer block in (say) `test/1.4/misc/T_unit_test_dmllib.dml`.

In particular, this could also act as a nice "documentation by example" for the contracts of these functions.